kate/grid/src/grid.rs

```diff
@@ -0,0 +1,294 @@
+		&self.inner
+	}
+}
+
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+impl<A: Clone + Send + Sync> RowMajor<A> {
+	pub fn row(&self, y: usize) -> Option<&[A]> {
+		if y >= self.height() {
+			return None;
+		}
+		let start = y.checked_mul(self.width())?;
+		let end = y.checked_add(1)?.checked_mul(self.width())?;
+		Some(&self.inner[start..end])
+	}
+
+	pub fn iter_col(&self, x: usize) -> Option<impl Iterator<Item = &A> + '_> {
+		if x >= self.width() {
+			return None;
+		}
+		Some((0..self.height()).map(move |y| self.get(x, y).expect("Bounds already checked")))
+	}
+
+	pub fn rows(&self) -> impl Iterator<Item = (usize, &[A])> + '_ {
+		(0..self.height()).map(|y| (y, self.row(y).expect("Bounds already checked")))
+	}
+
+	#[cfg(feature = "parallel")]
+	pub fn rows_par_iter(&self) -> impl ParallelIterator<Item = (usize, &[A])> + '_ {
+		(0..self.height())
+			.into_par_iter()
+			.map(|y| (y, self.row(y).expect("Bounds already checked")))
+	}
+
+	// TODO: this return type is kinda gross, should it just iterate over vecs?
+	pub fn columns(&self) -> impl Iterator<Item = (usize, impl Iterator<Item = &A>)> + '_ {
+		(0..self.width()).map(|x| (x, self.iter_col(x).expect("Bounds already checked")))
+	}
+
+	pub fn iter_row_wise(&self) -> impl Iterator<Item = &A> + '_ {
+		(0..self.height()).flat_map(move |y| {
+			(0..self.width()).map(move |x| self.get(x, y).expect("Bounds already checked"))
+		})
+	}
+
+	pub fn iter_column_wise(&self) -> impl Iterator<Item = &A> + '_ {
+		(0..self.width()).flat_map(move |x| {
+			(0..self.height()).map(move |y| self.get(x, y).expect("Bounds already checked"))
+		})
+	}
+
+	pub fn to_column_major(&self) -> ColumnMajor<A> {
+		self.iter_column_wise()
+			.map(Clone::clone)
+			.collect::<Vec<_>>()
+			.into_column_major(self.width(), self.height())
+			.expect("Bounds already checked")
+	}
+}
+
+impl<A: Clone> ColumnMajor<A> {
+	pub fn col(&self, x: usize) -> Option<&[A]> {
+		if x >= self.width() {
+			return None;
+		}
+		let start = x.checked_mul(self.height())?;
+		let end = x.checked_add(1)?.checked_mul(self.height())?;
+		Some(&self.inner[start..end])
+	}
+
+	pub fn iter_row(&self, y: usize) -> Option<impl Iterator<Item = &A> + '_> {
+		if y >= self.height() {
+			return None;
+		}
+		Some((0..self.width()).map(move |x| self.get(x, y).expect("Size checked at instantiation")))
+	}
+
+	pub fn iter_row_wise(&self) -> impl Iterator<Item = &A> + '_ {
+		(0..self.height()).flat_map(move |y| {
+			(0..self.width()).map(move |x| self.get(x, y).expect("Bounds already checked"))
+		})
+	}
+
+	pub fn iter_column_wise(&self) -> impl Iterator<Item = &A> + '_ {
+		(0..self.width()).flat_map(move |x| {
+			(0..self.height()).map(move |y| self.get(x, y).expect("Bounds already checked"))
+		})
+	}
+
+	pub fn to_row_major(&self) -> RowMajor<A> {
+		self.iter_row_wise()
+			.map(Clone::clone)
+			.collect::<Vec<_>>()
+			.into_row_major(self.width(), self.height())
+			.expect("Bounds already checked")
+	}
+}
```

Comment:
This doesn't really fall into the description of `Into`, where you are converting between two types that don't have wildly different data descriptions: for example `String` and `Vec<u8>`. Here, because we completely re-order all the elements of the vec in a way that does change the data layout, I would rather keep separate. Otherwise, if someone has a row wise type and is passing it to a column-wise method, they may unknowingly reallocate a huge amount of data.

Reply:
I'll add docs to make this clearer.